test/e2e/storage/testsuites/snapshottable.go

```diff
@@ -105,141 +109,275 @@ func (s *snapshottableTestSuite) DefineTests(driver TestDriver, pattern testpatt
-		framework.ExpectNoError(err)
+		init(l)
+		defer cleanup(l)
 
-		ginkgo.By("checking the snapshot")
-		// Get new copy of the snapshot
-		snapshot, err = dc.Resource(SnapshotGVR).Namespace(snapshot.GetNamespace()).Get(context.TODO(), snapshot.GetName(), metav1.GetOptions{})
-		framework.ExpectNoError(err)
+		TestSnapshottable(l, SnapshotClassTest{
+			DeletionPolicy: "Delete",
+		})
+		TestSnapshotDeleted(l, SnapshotClassTest{
+			DeletionPolicy: "Delete",
+		})
 
-		// Get the bound snapshotContent
-		snapshotStatus := snapshot.Object["status"].(map[string]interface{})
-		snapshotContentName := snapshotStatus["boundVolumeSnapshotContentName"].(string)
-		snapshotContent, err := dc.Resource(SnapshotContentGVR).Get(context.TODO(), snapshotContentName, metav1.GetOptions{})
-		framework.ExpectNoError(err)
+	})
 
-		snapshotContentSpec := snapshotContent.Object["spec"].(map[string]interface{})
-		volumeSnapshotRef := snapshotContentSpec["volumeSnapshotRef"].(map[string]interface{})
+	ginkgo.It("should not delete snapshot with retain policy [Feature:VolumeSnapshotDataSource]", func() {
+		l := &snapshottableLocal{}
 
-		// Check SnapshotContent properties
-		ginkgo.By("checking the SnapshotContent")
-		framework.ExpectEqual(snapshotContentSpec["volumeSnapshotClassName"], vsc.GetName())
-		framework.ExpectEqual(volumeSnapshotRef["name"], snapshot.GetName())
-		framework.ExpectEqual(volumeSnapshotRef["namespace"], snapshot.GetNamespace())
+		init(l)
+		defer cleanup(l)
+
+		TestSnapshottable(l, SnapshotClassTest{
+			DeletionPolicy: "Retain",
+		})
+		TestSnapshotDeleted(l, SnapshotClassTest{
+			DeletionPolicy: "Retain",
+		})
 	})
 }
 
+// snapshottableLocal is used to keep the current state of a snapshottable
+// test, associated objects, and cleanup steps.
+type snapshottableLocal struct {
+	config        *PerTestConfig
+	driverCleanup func()
+	cleanupSteps  []func()
+
+	cs        clientset.Interface
+	dc        dynamic.Interface
+	pvc       *v1.PersistentVolumeClaim
+	sc        *storagev1.StorageClass
+	pod       *v1.Pod
+	vsc       *unstructured.Unstructured
+	vs        *unstructured.Unstructured
+	vscontent *unstructured.Unstructured
+}
+
+// SnapshotClassTest represents parameters to be used by snapshot tests.
+// Not all parameters are used by all tests.
+type SnapshotClassTest struct {
+	DeletionPolicy string
+}
+
+// TestSnapshottable tests volume snapshots based on a given SnapshotClassTest
+func TestSnapshottable(l *snapshottableLocal, sct SnapshotClassTest) {
+	var err error
+
+	ginkgo.By("creating a SnapshotClass")
+	l.vsc.Object["deletionPolicy"] = sct.DeletionPolicy
+	l.vsc, err = l.dc.Resource(SnapshotClassGVR).Create(context.TODO(), l.vsc, metav1.CreateOptions{})
+	framework.ExpectNoError(err)
+	l.cleanupSteps = append(l.cleanupSteps, func() {
+		framework.Logf("deleting SnapshotClass %s", l.vsc.GetName())
+		l.dc.Resource(SnapshotClassGVR).Delete(context.TODO(), l.vsc.GetName(), metav1.DeleteOptions{})
+	})
```

Comment:
Lets keep it uniform as to what is added in cleanupsteps and what is deferred within  the function.  This can be a simple defer similar to line 266 where the snapshot is getting deleted.

Reply:
To me it looks like cleanup steps mainly cleanup things setup in the common init() call. resources setup inside the TestSnapshottable and TestSnapshotDeleted can be deleted by a defer inside the functions.